### PR TITLE
Events

### PR DESCRIPTION
**Pull request by skarzhevskyy**
1. Made TextBoxBase components work with focus and value change events as in browser.
   new method added InputElement.setNewValue  to simulate browser behavior with value change on before focus lost.
2. Changes to HTML parser, Fixed the original test .
   added toString() for element and nodes to better debugging experience in Eclipse

### CODE REVIEW

File: src/main/java/com/doctusoft/gwtmock/Document.java [hunk 1]
```diff
@@ -186,12 +197,23 @@ public void printFormatted(Element element, String indent, PrintWriter pw) {
 			pw.print(" " + attribute.getKey() + "=\"" + attribute.getValue() + "\"");
 		}
 		pw.println(">");
-		if (!Strings.isNullOrEmpty(element.getInnerText())) {
```
Comment (gabor-farkas) [inline src/main/java/com/doctusoft/gwtmock/Document.java:189]: why is this removed?

File: src/main/java/com/google/gwt/core/client/JavaScriptObject.java [hunk 2]
```diff
@@ -144,7 +146,11 @@ public native String toSource() /*-{
    */
   @Override
   public final String toString() {
-    return JavaScriptObject.class.desiredAssertionStatus() ?
-        toStringVerbose(this) : toStringSimple(this);
-  }
+      if (true) {
+          return Document.Instance.toStringVerbose(this);
+      } else {
+          return JavaScriptObject.class.desiredAssertionStatus() ?
```
Comment (gabor-farkas) [inline src/main/java/com/google/gwt/core/client/JavaScriptObject.java:152]: (I think we can just remove or comment previous gwt code)

File: src/main/java/com/google/gwt/dom/client/DOMImpl.java [hunk 3]
```diff
@@ -19,15 +19,108 @@
...
+                    fireValueChangeIfAny((InputElement)activeElement);
+                }
+                activeElement.fireEvent(BlurEvent.getType());
+            }
+            element.fireEvent(FocusEvent.getType());
+            activeElement = element;
+        }
+    }
+	
+    private boolean isAtached(Element elem) {
```
Comment (gabor-farkas) [inline src/main/java/com/google/gwt/dom/client/DOMImpl.java]: 'attached'

File: src/main/java/com/google/gwt/dom/client/InputElement.java [hunk 4]
```diff
@@ -381,6 +384,19 @@ public final void setValue(String value) {
...
   /**
+   * @Mock
+   * 
+   * Set value in native element (like Selenium 'sendKeys').
+   * The new value will be available in calls to getValue().
+   * N.B. Will not trigger ValueChange events until focus lost.
+   */
+  public final void setNewValue(String value) {
+      focus();
+      setAttribute("mock.Value.new", value);
```
Comment (gabor-farkas) [inline src/main/java/com/google/gwt/dom/client/InputElement.java]: this method seems to be specific to your testing habits, I think it would be better to keep this as a static helper in your project

File: src/main/java/com/doctusoft/gwtmock/Document.java (hunk 1 above)
Comment (gabor-farkas) [inline src/main/java/com/doctusoft/gwtmock/Document.java:189]: ahh, I see, the Text node is already handled, right

Comment (skarzhevskyy) [inline src/main/java/com/doctusoft/gwtmock/Document.java:189]: Yes. The text nodes are now nodes in Simulated DOM so no special handling is required.

e.g.  setInnerText(text) does this:  .appendChild(document.createTextNode(text));

File: src/main/java/com/google/gwt/core/client/JavaScriptObject.java (hunk 2 above)
Comment (skarzhevskyy) [inline src/main/java/com/google/gwt/core/client/JavaScriptObject.java:152]: Having reference to original code helps to understand differences in Mock and real implementation.
Also there is a still possibility  that I will do the same exercise for GWT 2.8.

File: src/main/java/com/google/gwt/dom/client/DOMImpl.java [hunk 5]
```diff
@@ -19,15 +19,108 @@
...
+            element.fireEvent(FocusEvent.getType());
+            activeElement = element;
+        }
+    }
+
+    private boolean isAttached(Element elem) {
+        return DOM.getEventListener(elem) != null;
+    }
+
+    private void fireValueChangeIfAny(InputElement element) {
```
Comment (skarzhevskyy) [inline src/main/java/com/google/gwt/dom/client/DOMImpl.java]: @gabor-farkas  see this. Regarding InputElement.setNewValue

File: src/main/java/com/google/gwt/dom/client/InputElement.java (hunk 4 above)
Comment (skarzhevskyy) [inline src/main/java/com/google/gwt/dom/client/InputElement.java]: In my project I do have a static call to this  'GwtMock.type(debugId, text)'.
And this function is not alone. It make the function DOMImpl.fireValueChangeIfAny actually work. 

Also there are Unit tests for this in this library. See class TestTextBoxEvents.
This is make the library usable to testing events on TextBoxBase components.
The are firing change events only on focus lost.

But agree that this function better belong in some other place in this library that we don't have yet.   Something like   BrowserMock  or GWTMockBridge. I will open a special discussion for this.

Also the function name is not perfect.  I wanted to runt this by you in hope that we will come up with more usable name and approach.

---

Also I that believe order of Change and FosusLost events in browser may happens to be different that one we are expecting (according to spec) in application.  As described here  http://stackoverflow.com/questions/1165522/javascript-events-not-firing-in-the-expected-order
I did observed this in production and in QA of my system.

This is actually I wanted later to add more Mock to such cases to make my library more resilient ...

File: src/main/java/com/google/gwt/core/client/JavaScriptObject.java (hunk 2 above)
Comment (gabor-farkas) [inline src/main/java/com/google/gwt/core/client/JavaScriptObject.java:152]: We can also keep the original code in a comment

File: src/main/java/com/google/gwt/core/shared/GWT.java [hunk 6]
```diff
@@ -62,17 +67,82 @@
 		 */
 		public Object create(Class<?> classLiteral);
 	}
-	
+
+    private static class SimplifiedGWTCreateSupplier implements CustomGWTCreateSupplier {
```
Comment (gabor-farkas) [inline src/main/java/com/google/gwt/core/shared/GWT.java]: `ClassToInstanceGWTCreateSupplier` would be a more talking name I think

File: src/main/java/com/google/gwt/core/shared/GWT.java [hunk 7]
```diff
@@ -62,17 +67,82 @@
...
+	
+	private static SimplifiedGWTCreateSupplier mockSupplier = createDefaultGWTCreateSupplier();
+	
 	public static void addCustomSupplier(CustomGWTCreateSupplier customSupplier) {
 		customSuppliers.add(customSupplier);
+		simplifiedGWTCreateCustomSupplier = null;
 	}
+
+    public static void registerGWTCreateImplementation(Class<?> classLiteral, Class<?> classLiteralImplementation) {
+        if (simplifiedGWTCreateCustomSupplier == null) {
```
Comment (gabor-farkas) [inline src/main/java/com/google/gwt/core/shared/GWT.java]: shouldn't this method register the instance to `mockSupplier`? `simplifiedGWTCreateCustomSupplier` doesn't seem to be used anywhere.

Comment (gabor-farkas) [inline src/main/java/com/google/gwt/core/shared/GWT.java]: Ahh sorry, I see it's added using addCustomSupplier. But I don't get this, the addCustomSupplier method nulls this field out, why do you have the field?
You could just register the instance to `mockSupplier` directly I think.

Comment (skarzhevskyy) [inline src/main/java/com/google/gwt/core/shared/GWT.java]: I fixed the filed set to null. It was intended to have single instance of "customSupplierClassToInstance" (new name) and reset it to null upon tearDown()

mockSupplier is non re-settable default  implementation provided by this Library.

File: src/main/java/com/google/gwt/core/client/JavaScriptObject.java (hunk 2 above)
Comment (skarzhevskyy) [inline src/main/java/com/google/gwt/core/client/JavaScriptObject.java:152]: It is your call.
The codes inside `if (false) {...}` is not called but is searchable by Eclipse References Workplace. sometime it is useful .

**Comment by skarzhevskyy** [comment]
@gabor-farkas  Please let me know the status of this PR

I have other changes coming that are based on this branch.
[ListBox](https://github.com/skarzhevskyy/gwt-mock-2.5.1/tree/ListBox-mock)
[MenuBar](https://github.com/skarzhevskyy/gwt-mock-2.5.1/tree/MenuBar-mock)

that I don't want to create  PR yet until this is merged to separate the list of changes for review.

File: src/main/java/com/google/gwt/dom/client/DOMImpl.java (hunk 5 above)
Comment (gabor-farkas) [inline src/main/java/com/google/gwt/dom/client/DOMImpl.java]: Ok, I see. Nevertheless, I don't think we should start adding test utility/helper methods to core gwt classes. This scaffolding can work just the same in a static helper way.

In our project we also have static helpers in our common superclass for gwt mock testing: helper methods for clicking on various elements, or selecting a given value from a combobox and stuff like that.

File: src/main/java/com/google/gwt/dom/client/InputElement.java (hunk 4 above)
Comment (gabor-farkas) [inline src/main/java/com/google/gwt/dom/client/InputElement.java]: About testing Change and FocusLost events: if I understand things right, your application might accidentally rely on the order of Change and Blur events and you want to test that your application work fine in both cases (Change-Blur, and Blur-Change). To test this, I think you can just programmtically fire these events in both orders, and these don't need special extensions to gwt-mock. I would also add a reusable utility class that encapsulates this blur-change order problem and your components can rely on that class. What do you think?

File: src/main/java/com/google/gwt/dom/client/DOMImpl.java (hunk 5 above)
Comment (skarzhevskyy) [inline src/main/java/com/google/gwt/dom/client/DOMImpl.java]: Exactly this is what I want to avoid!  I want users of the library to be able to start using it right away without creating multiple helper classes. Now I also have large Helper class in my project.  I don't believe it is good approach to force users to create  helper classes.   On another hand they should be able to extend the helper classes but this is different story.

So I see the current implementation .  I see you have DOMImpl.buttonClick.
How is my additions are different ?

File: src/main/java/com/google/gwt/dom/client/InputElement.java (hunk 4 above)
Comment (skarzhevskyy) [inline src/main/java/com/google/gwt/dom/client/InputElement.java]: I indeed to use this library for UI components testing.  eventually  I will move my components for JS implementations and will just test controller using this library. Sorry but I don't see the complete path yet, I just hope that I will be able to so.

So I think the problem I see is maybe common pattern of order of events handling. So other people will reuse the same approach or will make it better.
For now I'm not suggesting any implementation to this mock library for this, But I intend to do so soon once I do some research and find how to other people are testing the same.

File: src/main/java/com/google/gwt/dom/client/DOMImpl.java (hunk 5 above)
Comment (gabor-farkas) [inline src/main/java/com/google/gwt/dom/client/DOMImpl.java]: DOMImpl.buttonClick is part of the original GWT source, and it's used by ButtonElement.click(), that's why it has to be implemented.

I think it's a good approach to keep static helpers separately. It's the single responsibility principle. You don't bundle things up to a big monolith. Other developers might need different helpers or other users would want a different API for their helpers. You can start a gwt-mock-helpers project, that hosts the helpers we commonly use, and then others can use that too. But these helpers are definitiely not part of gwt-mock.

Comment (skarzhevskyy) [inline src/main/java/com/google/gwt/dom/client/DOMImpl.java]: I see you have a very solid point. I wish I new it from the start.

I'm very hesitant to create an maintain a new project ' gwt-mock-helpers'.  Can it be a part of this project as a different sub-module or a new Class?
 The problem is that in many case it will rely anyway on mock implementation and changers to 'gwt-mock-2.5.1'  will be required at the same time.

## 

I will try to start all over in a new PR.

File: src/main/java/com/google/gwt/dom/client/DOMImpl.java (hunk 3 above)
Comment (skarzhevskyy) [inline src/main/java/com/google/gwt/dom/client/DOMImpl.java]: Gábor
  Now analyzing and redoing the events implementation in GwtMock I see what
caused the major confusion for me to implement it DOMImpl and in Element.

  You have function Element.fireEvent  if we are to stick to the rules "keep
static helpers separately."   Need to move off the GWT code to avoid other
developers doing the same mistake as I did.

Regards,
Vlad

## 

Vlad

Comment (gabor-farkas) [inline src/main/java/com/google/gwt/dom/client/DOMImpl.java]: I think there's a slight but important difference: fireEvent on element is just GWT-side pair of Element.dispatchEvent, and as __listener and __eventbits are also naturally there on the Element (as they are there internally in a real browser too), the fireEvent method is also naturally there. So it's rather a part of the 'browser-simulation' than a test utility.
